core/src/main/java/org/jruby/util/Dir.java

```diff
@@ -973,16 +973,24 @@ private static int glob_helper(Ruby runtime, String cwd, byte[] scheme,
                         }
                         if ( fnmatch(magic, 0, magic.length, fileBytes, 0, fileBytes.length, flags) == 0 ) {
                             buf.length(0);
-                            if (scheme != null && SLASH_INDEX == -1) buf.append(scheme);
                             buf.append(base);
                             buf.append( isRoot(base) ? EMPTY : SLASH );
                             buf.append( getBytesInUTF8(file) );
-                            if ( SLASH_INDEX == -1 ) {
+                            resource = JRubyFile.createResource(runtime, cwd, RubyString.byteListToString(buf));
```

Comment:
Up above we see the same resource creation but it uses the encoding for charset:

```java
resource = JRubyFile.createResource(runtime, cwd, new String(buf.unsafeBytes(), buf.begin(), buf.length(), enc.getCharset()));
```

This enc comes from the original pattern and that pattern is fed through RubyFile#get_path/filePathConvert which has special logic for transcoding if NOT on windows to match filesystem encoding.   I think the use of byteListToString (used both here and below) is wrong.  It will make an ISO-8859-1 string where each byte is a char vs the line above which properly decodes the bytes given with the charset provided.

I think the reason we do not ever see problems is almost all files are single-byte ASCII.

Reply:
Yes I think this risks losing encoding and possibly dealing with mangled characters. If we cannot fix it by working with the possibly-multibyte characters directly, then we should use the functions in RubyEncoding to decode them to String properly.

Reply:
sorted